nucleus/hk2/hk2-config/src/main/java/org/jvnet/hk2/config/Dom.java

```diff
@@ -328,8 +328,10 @@ public ServiceLocator getHabitat() {
     public Dom(Dom source, Dom parent) {
         this(source.getHabitat(), source.document, parent, source.model);
         List<Child> newChildren = new ArrayList<Child>();
-        for (Child child : source.children) {
-            newChildren.add(child.deepCopy(this));
+        synchronized (source) {
```

Comment:
https://rules.sonarsource.com/java/tag/multi-threading/RSPEC-2445

Reply:
Synchronized blocks of non-final fields are not necessarily a problem.
Unlike the URL example, this constructor does not change "source.children".

Reply:
> Unlike the URL example, 

It'd be extremely interesting and less fun should the example considered this `Dom.java` exactly with that change.

> this constructor does not change "source.children".

Can `source.children` be changed externally to this Ctor?
Can it happen in different thread in not-synchronized (on the very same `source` object) way?
Can it be controlled from this point how `source` is used in other threads/places?

Reply:
> Can `source.children` be changed externally to this Ctor?

So I was wrong in this part, as access to `children` is synchronized on `source`. 

> Can it happen in different thread in not-synchronized (on the very same source object) way?

Probably it can not (now).

This change indeed prevents `ConcurrentModificationException` from happening, as I saw in my unit test.  (CMEx happens in previous `Dom` version).

Sorry for the noise!